File: 2025/12/05/cell-based-architecture/src/middleware.ts
import { Request, Response, NextFunction } from 'express';
import { CellRouter } from './router';
import { CellContext } from './types';

declare global {
  namespace Express {
    interface Request {
      cellContext?: CellContext;
    }
  }
}

export function cellAwareMiddleware(router: CellRouter) {
  return async (req: Request, res: Response, next: NextFunction) => {
    try {
      // Extract tenant ID from auth token or header
      const tenantId = extractTenantId(req);
      if (!tenantId) {
        return res.status(401).json({ error: 'Missing tenant ID' });
      }

      // Look up cell ID
      const cellId = await router.getCellForTenant(tenantId);
      if (!cellId) {
        return res.status(503).json({ 
          error: 'No cell available for tenant',
          tenantId 
        });
      }

      // Attach cell context to request
      req.cellContext = {
        tenantId,
        cellId,
        region: extractRegion(req),
      };

      // Add cell ID to header for downstream services
      req.headers['x-cell-id'] = cellId;
      req.headers['x-tenant-id'] = tenantId;

      next();
    } catch (error) {
      console.error('Cell routing error:', error);
      res.status(500).json({ error: 'Internal routing error' });
    }
  };
}

function extractTenantId(req: Request): string | null {
  // Try JWT token first
  const authHeader = req.headers.authorization;
  if (authHeader?.startsWith('Bearer ')) {
    const token = authHeader.substring(7);
    const payload = parseJWT(token);
    if (payload?.tenantId) {
      return payload.tenantId;
    }
  }

  // Fallback to header
  return (req.headers['x-tenant-id'] as string) || null;
}

function extractRegion(req: Request): string | undefined {
  return (req.headers['x-region'] as string) || 
         (req.headers['cf-ipcountry'] as string) || 
         undefined;
}

function parseJWT(token: string): any {
  try {
    const parts = token.split('.');
    if (parts.length !== 3) {
      return null;
    }
    const payload = JSON.parse(Buffer.from(parts[1], 'base64').toString());
    return payload;
  } catch {
    return null;
  }
}

File: 2025/12/05/cell-based-architecture/go/middleware.go
package main

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
)

// CellContext contains cell routing information
type CellContext struct {
	TenantID string
	CellID   string
	Region   string
}

type contextKey string

const cellContextKey contextKey = "cellContext"

// CellAwareMiddleware creates middleware that routes requests to the correct cell
func CellAwareMiddleware(router CellRouter) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Extract tenant ID
			tenantID := extractTenantID(r)
			if tenantID == "" {
				http.Error(w, `{"error":"Missing tenant ID"}`, http.StatusUnauthorized)
				return
			}

			// Look up cell ID
			cellID, err := router.GetCellForTenant(tenantID)
			if err != nil {
				http.Error(w, fmt.Sprintf(`{"error":"No cell available for tenant","tenantId":"%s"}`, tenantID), http.StatusServiceUnavailable)
				return
			}

			// Create cell context
			cellContext := CellContext{
				TenantID: tenantID,
				CellID:   cellID,
				Region:   extractRegion(r),
			}

			// Add to request context
			ctx := context.WithValue(r.Context(), cellContextKey, cellContext)
			r = r.WithContext(ctx)

			// Add headers for downstream services
			r.Header.Set("X-Cell-ID", cellID)
			r.Header.Set("X-Tenant-ID", tenantID)

			next.ServeHTTP(w, r)
		})
	}
}

// GetCellContext extracts cell context from request
func GetCellContext(r *http.Request) *CellContext {
	ctx := r.Context().Value(cellContextKey)
	if ctx == nil {
		return nil
	}
	cellCtx, ok := ctx.(CellContext)
	if !ok {
		return nil
	}
	return &cellCtx
}

func extractTenantID(r *http.Request) string {
	// Try JWT token first
	authHeader := r.Header.Get("Authorization")
	if strings.HasPrefix(authHeader, "Bearer ") {
		token := strings.TrimPrefix(authHeader, "Bearer ")
		if tenantID := parseJWT(token); tenantID != "" {
			return tenantID
		}
	}

	// Fallback to header
	return r.Header.Get("X-Tenant-ID")
}

func extractRegion(r *http.Request) string {
	if region := r.Header.Get("X-Region"); region != "" {
		return region
	}
	if region := r.Header.Get("Cf-Ipcountry"); region != "" {
		return region
	}
	return ""
}

func parseJWT(token string) string {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return ""
	}

	payload, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return ""
	}

	var claims map[string]interface{}
	if err := json.Unmarshal(payload, &claims); err != nil {
		return ""
	}

	if tenantID, ok := claims["tenantId"].(string); ok {
		return tenantID
	}

	return ""
}
